scylla-macros/src/from_user_type.rs

```diff
@@ -36,12 +36,12 @@ pub fn from_user_type_derive(tokens_input: TokenStream) -> TokenStream {
         impl FromCQLVal<scylla::frame::response::result::CQLValue> for #struct_name {
             fn from_cql(cql_val: scylla::frame::response::result::CQLValue)
             -> Result<Self, scylla::cql_to_rust::FromCQLValError> {
-                use std::collections::HashMap;
+                use std::collections::BTreeMap;
```

Comment:
@Kejmer 
I know it was done for some reason, I cannot recall now why  - please explain.

Reply:
Having HashMap as a key value requires it to has a Hash trait implemented and HashMap doesn't provide default implementation and BTreeMap provides the same functionality and has default implementation for Hash.

Reply:
@psarna

Reply:
@Kejmer where exactly do you need to use HashMap as a key? Can you point me to the line in code? BTreeMap may provide similar functionality, but does not have identical performance characteristics.

**If** we want to migrate from HashMap to BTreeMap, then it cannot be done in the same commit which implements CQL types - it's an unrelated change. But for starters I need to understand why we're interested in storing the whole map as a key - while BTreeMap provides a Hash implementation, it's likely O(n), so it would generally be suboptimal to use it as a key anywhere. Also, we don't need ordering of the fields, so hash map is simply a better choice performance-wise, both in terms of execution time and memory utilization.

Reply:
@psarna 
result.rs:54 has #[derive(Hash)]
cql_to_rust.rs:87 requires first type to be hashable

Since Map accepts any CQLType as key value, it could be a Map as well.

Reply:
Ok, then I'll rather provide a hasher for this hash map (in a separate commit) than change the underlying type. I'm not 100% convinced, since it might be premature optimization, but on the other hand UDTs are themselves an optimization, so we might want to keep their implementation faster if the price is low - implementing a single hasher. But first, I need a second opinion :) @piodul?

Reply:
I'm trying to break this commit into a bunch of smaller ones.

Reply:
@psarna if i remember correctly there is also a problem with implementing Hash trait to HashMap since neither trait or class was declared in our files and one workaround we found was to create a class wrapper that contains only HashMap inside and then implement a Hash function for it, but there is still problem of HashMap being unordered, so  two maps with same keys and same elements can return different Hash values and as a result create future bugs.

Reply:
That's a valid concern (although it's still possible to compute a hash which disregards the order of elements). I'm generally fine with using BTreeMap instead, since there is a sane reasoning behind the change, but please list all these arguments in the commit message of the separate commit which changes the map type.

Reply:
@psarna 
Would it be a problem if I create another PR for this task and close this one? 
I've cherry-picked this one big commit and broke it down into smaller ones, it was easier for me that way.

Reply:
No problem, although you can also simply force-push to this PR's branch, and the git history will get rewritten.

Reply:
(force-push == `git push -f`, or, better yet, `git push --force-with-lease`)

Reply:
@psarna I merged cherry-picked branch with ours strategy. Is everything ok now in your opinion?

Reply:
I have no idea what happened to git history here - we generally want to avoid merges from inside the pull request. The rule of thumb is that if you want to type `git merge` in our repo - don't :) Merging branches is left for the maintainers only, so that we sustain a clear, forward history of changes. Same goes with `git pull`, unless you use `git pull --rebase`.

Please repush without this merge commit. You can do it e.g. like this:
1. Checkout main: `git fetch; git checkout origin/main`
2. Cherry-pick your commits on top of main: `git cherry-pick 351afdf02 14776f4d8a f7e5284144 7247b4e74 c209b2a0c 41c70d79e d30712861a d30712861a`
3. Force-push

Reply:
This method doesn't work, I tried to `git push --force-with-lease origin add_collecions` but it doesn't overwrite changes. I think creating new PR would be a less painfull way with clear commit history.

Reply:
fine with me